In [8]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import os
print(os.listdir("cell_images/"))
from keras.models import Sequential
ORIGIN_PATH=os.getcwd()
DATA_ORIGIN_PATH = os.path.abspath("cell_images")
TRAINING_GENERATOR_PATH = os.path.abspath("MalariaTrainData")
TESTING_GENERATOR_PATH = os.path.abspath("MalariaTestData")


['Parasitized', 'Uninfected']


# Data Restructuring
I couldn't find a straightforward method to separate the data into Train/Test. Keras ImageDataGenerator provides the means to splitting the data into training/validation, yet I'd like to test the model on a data that it hasn't seen before (not even through validation)

The following few celles creates 2 folders, containing Training data and Testing data. The images change randomly each run.

In [9]:
from keras.preprocessing.image import ImageDataGenerator

try :
    os.chdir(ORIGIN_PATH)
    os.mkdir("MalariaTrainData")
    os.chdir("MalariaTrainData")
    os.mkdir("Parasitized")
    os.mkdir("Uninfected")
    os.chdir(ORIGIN_PATH)
    os.mkdir("MalariaTestData")
    os.chdir("MalariaTestData")
    os.mkdir("Parasitized")
    os.mkdir("Uninfected")
    os.chdir(ORIGIN_PATH)

except :
    pass

In [10]:
test_precentage = 0.3


destination_1_path_train = os.path.join(ORIGIN_PATH,"MalariaTrainData","Parasitized")
destination_0_path_train = os.path.join(ORIGIN_PATH,"MalariaTrainData","Uninfected")
destination_1_path_test = os.path.join(ORIGIN_PATH,"MalariaTestData","Parasitized")
destination_0_path_test = os.path.join(ORIGIN_PATH,"MalariaTestData","Uninfected")
source_1_path = os.path.join(DATA_ORIGIN_PATH,"Parasitized")
source_0_path = os.path.join(DATA_ORIGIN_PATH,"Uninfected")
import random
from shutil import copyfile

destination_1_path=""
destination_0_path=""

os.chdir(source_1_path)
imgs = os.listdir()
total = len(imgs)
progress=0.0
for img in imgs:
    progress+=1
    file_path = os.path.join(source_1_path,img)
    if random.random() > (1-test_precentage):
        destination_1_path = destination_1_path_test
    else :
        destination_1_path = destination_1_path_train
        
    dest_path = os.path.join(destination_1_path,img)
    copyfile(file_path,dest_path)
    print('%.3f'%(100*progress/total))
    
os.chdir(source_0_path)
imgs = os.listdir()
total = len(imgs)
progress=0.0
for img in imgs:
    progress+=1
    file_path = os.path.join(source_0_path,img)
    if random.random() > (1-test_precentage):
        destination_0_path = destination_0_path_test
    else :
        destination_0_path = destination_0_path_train
        
    dest_path = os.path.join(destination_0_path,img)
    copyfile(file_path,dest_path)
    print('%.3f'%(100*progress/total))
os.chdir(ORIGIN_PATH)

0.007
0.015
0.022
0.029
0.036
0.044
0.051
0.058
0.065
0.073
0.080
0.087
0.094
0.102
0.109
0.116
0.123
0.131
0.138
0.145
0.152
0.160
0.167
0.174
0.181
0.189
0.196
0.203
0.210
0.218
0.225
0.232
0.239
0.247
0.254
0.261
0.269
0.276
0.283
0.290
0.298
0.305
0.312
0.319
0.327
0.334
0.341
0.348
0.356
0.363
0.370
0.377
0.385
0.392
0.399
0.406
0.414
0.421
0.428
0.435
0.443
0.450
0.457
0.464
0.472
0.479
0.486
0.493
0.501
0.508
0.515
0.522
0.530
0.537
0.544
0.552
0.559
0.566
0.573
0.581
0.588
0.595
0.602
0.610
0.617
0.624
0.631
0.639
0.646
0.653
0.660
0.668
0.675
0.682
0.689
0.697
0.704
0.711
0.718
0.726
0.733
0.740
0.747
0.755
0.762
0.769
0.776
0.784
0.791
0.798
0.806
0.813
0.820
0.827
0.835
0.842
0.849
0.856
0.864
0.871
0.878
0.885
0.893
0.900
0.907
0.914
0.922
0.929
0.936
0.943
0.951
0.958
0.965
0.972
0.980
0.987
0.994
1.001
1.009
1.016
1.023
1.030
1.038
1.045
1.052
1.060
1.067
1.074
1.081
1.089
1.096
1.103
1.110
1.118
1.125
1.132
1.139
1.147
1.154
1.161
1.168
1.176
1.183
1.190
1.197
1.205
1.21

9.956
9.964
9.971
9.978
9.985
9.993
10.000
10.007
10.015
10.022
10.029
10.036
10.044
10.051
10.058
10.065
10.073
10.080
10.087
10.094
10.102
10.109
10.116
10.123
10.131
10.138
10.145
10.152
10.160
10.167
10.174
10.181
10.189
10.196
10.203
10.210
10.218
10.225
10.232
10.239
10.247
10.254
10.261
10.269
10.276
10.283
10.290
10.298
10.305
10.312
10.319
10.327
10.334
10.341
10.348
10.356
10.363
10.370
10.377
10.385
10.392
10.399
10.406
10.414
10.421
10.428
10.435
10.443
10.450
10.457
10.464
10.472
10.479
10.486
10.493
10.501
10.508
10.515
10.522
10.530
10.537
10.544
10.552
10.559
10.566
10.573
10.581
10.588
10.595
10.602
10.610
10.617
10.624
10.631
10.639
10.646
10.653
10.660
10.668
10.675
10.682
10.689
10.697
10.704
10.711
10.718
10.726
10.733
10.740
10.747
10.755
10.762
10.769
10.776
10.784
10.791
10.798
10.806
10.813
10.820
10.827
10.835
10.842
10.849
10.856
10.864
10.871
10.878
10.885
10.893
10.900
10.907
10.914
10.922
10.929
10.936
10.943
10.951
10.958
10.965
10.972
10.980
10.987
10.99

18.752
18.759
18.766
18.774
18.781
18.788
18.795
18.803
18.810
18.817
18.824
18.832
18.839
18.846
18.853
18.861
18.868
18.875
18.882
18.890
18.897
18.904
18.911
18.919
18.926
18.933
18.940
18.948
18.955
18.962
18.970
18.977
18.984
18.991
18.999
19.006
19.013
19.020
19.028
19.035
19.042
19.049
19.057
19.064
19.071
19.078
19.086
19.093
19.100
19.107
19.115
19.122
19.129
19.136
19.144
19.151
19.158
19.165
19.173
19.180
19.187
19.194
19.202
19.209
19.216
19.224
19.231
19.238
19.245
19.253
19.260
19.267
19.274
19.282
19.289
19.296
19.303
19.311
19.318
19.325
19.332
19.340
19.347
19.354
19.361
19.369
19.376
19.383
19.390
19.398
19.405
19.412
19.419
19.427
19.434
19.441
19.448
19.456
19.463
19.470
19.478
19.485
19.492
19.499
19.507
19.514
19.521
19.528
19.536
19.543
19.550
19.557
19.565
19.572
19.579
19.586
19.594
19.601
19.608
19.615
19.623
19.630
19.637
19.644
19.652
19.659
19.666
19.673
19.681
19.688
19.695
19.702
19.710
19.717
19.724
19.731
19.739
19.746
19.753
19.761
19.768
19.775
19.782

27.612
27.620
27.627
27.634
27.642
27.649
27.656
27.663
27.671
27.678
27.685
27.692
27.700
27.707
27.714
27.721
27.729
27.736
27.743
27.750
27.758
27.765
27.772
27.779
27.787
27.794
27.801
27.808
27.816
27.823
27.830
27.837
27.845
27.852
27.859
27.866
27.874
27.881
27.888
27.896
27.903
27.910
27.917
27.925
27.932
27.939
27.946
27.954
27.961
27.968
27.975
27.983
27.990
27.997
28.004
28.012
28.019
28.026
28.033
28.041
28.048
28.055
28.062
28.070
28.077
28.084
28.091
28.099
28.106
28.113
28.120
28.128
28.135
28.142
28.149
28.157
28.164
28.171
28.179
28.186
28.193
28.200
28.208
28.215
28.222
28.229
28.237
28.244
28.251
28.258
28.266
28.273
28.280
28.287
28.295
28.302
28.309
28.316
28.324
28.331
28.338
28.345
28.353
28.360
28.367
28.374
28.382
28.389
28.396
28.403
28.411
28.418
28.425
28.433
28.440
28.447
28.454
28.462
28.469
28.476
28.483
28.491
28.498
28.505
28.512
28.520
28.527
28.534
28.541
28.549
28.556
28.563
28.570
28.578
28.585
28.592
28.599
28.607
28.614
28.621
28.628
28.636
28.643

36.343
36.350
36.357
36.364
36.372
36.379
36.386
36.393
36.401
36.408
36.415
36.422
36.430
36.437
36.444
36.451
36.459
36.466
36.473
36.480
36.488
36.495
36.502
36.509
36.517
36.524
36.531
36.538
36.546
36.553
36.560
36.567
36.575
36.582
36.589
36.597
36.604
36.611
36.618
36.626
36.633
36.640
36.647
36.655
36.662
36.669
36.676
36.684
36.691
36.698
36.705
36.713
36.720
36.727
36.734
36.742
36.749
36.756
36.763
36.771
36.778
36.785
36.792
36.800
36.807
36.814
36.821
36.829
36.836
36.843
36.851
36.858
36.865
36.872
36.880
36.887
36.894
36.901
36.909
36.916
36.923
36.930
36.938
36.945
36.952
36.959
36.967
36.974
36.981
36.988
36.996
37.003
37.010
37.017
37.025
37.032
37.039
37.046
37.054
37.061
37.068
37.075
37.083
37.090
37.097
37.104
37.112
37.119
37.126
37.134
37.141
37.148
37.155
37.163
37.170
37.177
37.184
37.192
37.199
37.206
37.213
37.221
37.228
37.235
37.242
37.250
37.257
37.264
37.271
37.279
37.286
37.293
37.300
37.308
37.315
37.322
37.329
37.337
37.344
37.351
37.358
37.366
37.373

45.450
45.457
45.464
45.472
45.479
45.486
45.493
45.501
45.508
45.515
45.522
45.530
45.537
45.544
45.552
45.559
45.566
45.573
45.581
45.588
45.595
45.602
45.610
45.617
45.624
45.631
45.639
45.646
45.653
45.660
45.668
45.675
45.682
45.689
45.697
45.704
45.711
45.718
45.726
45.733
45.740
45.747
45.755
45.762
45.769
45.776
45.784
45.791
45.798
45.806
45.813
45.820
45.827
45.835
45.842
45.849
45.856
45.864
45.871
45.878
45.885
45.893
45.900
45.907
45.914
45.922
45.929
45.936
45.943
45.951
45.958
45.965
45.972
45.980
45.987
45.994
46.001
46.009
46.016
46.023
46.030
46.038
46.045
46.052
46.060
46.067
46.074
46.081
46.089
46.096
46.103
46.110
46.118
46.125
46.132
46.139
46.147
46.154
46.161
46.168
46.176
46.183
46.190
46.197
46.205
46.212
46.219
46.226
46.234
46.241
46.248
46.255
46.263
46.270
46.277
46.284
46.292
46.299
46.306
46.313
46.321
46.328
46.335
46.343
46.350
46.357
46.364
46.372
46.379
46.386
46.393
46.401
46.408
46.415
46.422
46.430
46.437
46.444
46.451
46.459
46.466
46.473
46.480

54.390
54.398
54.405
54.412
54.419
54.427
54.434
54.441
54.448
54.456
54.463
54.470
54.478
54.485
54.492
54.499
54.507
54.514
54.521
54.528
54.536
54.543
54.550
54.557
54.565
54.572
54.579
54.586
54.594
54.601
54.608
54.615
54.623
54.630
54.637
54.644
54.652
54.659
54.666
54.673
54.681
54.688
54.695
54.702
54.710
54.717
54.724
54.731
54.739
54.746
54.753
54.761
54.768
54.775
54.782
54.790
54.797
54.804
54.811
54.819
54.826
54.833
54.840
54.848
54.855
54.862
54.869
54.877
54.884
54.891
54.898
54.906
54.913
54.920
54.927
54.935
54.942
54.949
54.956
54.964
54.971
54.978
54.985
54.993
55.000
55.007
55.015
55.022
55.029
55.036
55.044
55.051
55.058
55.065
55.073
55.080
55.087
55.094
55.102
55.109
55.116
55.123
55.131
55.138
55.145
55.152
55.160
55.167
55.174
55.181
55.189
55.196
55.203
55.210
55.218
55.225
55.232
55.239
55.247
55.254
55.261
55.269
55.276
55.283
55.290
55.298
55.305
55.312
55.319
55.327
55.334
55.341
55.348
55.356
55.363
55.370
55.377
55.385
55.392
55.399
55.406
55.414
55.421

63.396
63.403
63.411
63.418
63.425
63.433
63.440
63.447
63.454
63.462
63.469
63.476
63.483
63.491
63.498
63.505
63.512
63.520
63.527
63.534
63.541
63.549
63.556
63.563
63.570
63.578
63.585
63.592
63.599
63.607
63.614
63.621
63.628
63.636
63.643
63.650
63.657
63.665
63.672
63.679
63.687
63.694
63.701
63.708
63.716
63.723
63.730
63.737
63.745
63.752
63.759
63.766
63.774
63.781
63.788
63.795
63.803
63.810
63.817
63.824
63.832
63.839
63.846
63.853
63.861
63.868
63.875
63.882
63.890
63.897
63.904
63.911
63.919
63.926
63.933
63.940
63.948
63.955
63.962
63.970
63.977
63.984
63.991
63.999
64.006
64.013
64.020
64.028
64.035
64.042
64.049
64.057
64.064
64.071
64.078
64.086
64.093
64.100
64.107
64.115
64.122
64.129
64.136
64.144
64.151
64.158
64.165
64.173
64.180
64.187
64.194
64.202
64.209
64.216
64.224
64.231
64.238
64.245
64.253
64.260
64.267
64.274
64.282
64.289
64.296
64.303
64.311
64.318
64.325
64.332
64.340
64.347
64.354
64.361
64.369
64.376
64.383
64.390
64.398
64.405
64.412
64.419
64.427

72.003
72.010
72.017
72.025
72.032
72.039
72.046
72.054
72.061
72.068
72.075
72.083
72.090
72.097
72.104
72.112
72.119
72.126
72.134
72.141
72.148
72.155
72.163
72.170
72.177
72.184
72.192
72.199
72.206
72.213
72.221
72.228
72.235
72.242
72.250
72.257
72.264
72.271
72.279
72.286
72.293
72.300
72.308
72.315
72.322
72.329
72.337
72.344
72.351
72.358
72.366
72.373
72.380
72.388
72.395
72.402
72.409
72.417
72.424
72.431
72.438
72.446
72.453
72.460
72.467
72.475
72.482
72.489
72.496
72.504
72.511
72.518
72.525
72.533
72.540
72.547
72.554
72.562
72.569
72.576
72.583
72.591
72.598
72.605
72.612
72.620
72.627
72.634
72.642
72.649
72.656
72.663
72.671
72.678
72.685
72.692
72.700
72.707
72.714
72.721
72.729
72.736
72.743
72.750
72.758
72.765
72.772
72.779
72.787
72.794
72.801
72.808
72.816
72.823
72.830
72.837
72.845
72.852
72.859
72.866
72.874
72.881
72.888
72.896
72.903
72.910
72.917
72.925
72.932
72.939
72.946
72.954
72.961
72.968
72.975
72.983
72.990
72.997
73.004
73.012
73.019
73.026
73.033

80.617
80.624
80.631
80.639
80.646
80.653
80.660
80.668
80.675
80.682
80.689
80.697
80.704
80.711
80.718
80.726
80.733
80.740
80.747
80.755
80.762
80.769
80.776
80.784
80.791
80.798
80.806
80.813
80.820
80.827
80.835
80.842
80.849
80.856
80.864
80.871
80.878
80.885
80.893
80.900
80.907
80.914
80.922
80.929
80.936
80.943
80.951
80.958
80.965
80.972
80.980
80.987
80.994
81.001
81.009
81.016
81.023
81.030
81.038
81.045
81.052
81.060
81.067
81.074
81.081
81.089
81.096
81.103
81.110
81.118
81.125
81.132
81.139
81.147
81.154
81.161
81.168
81.176
81.183
81.190
81.197
81.205
81.212
81.219
81.226
81.234
81.241
81.248
81.255
81.263
81.270
81.277
81.284
81.292
81.299
81.306
81.313
81.321
81.328
81.335
81.343
81.350
81.357
81.364
81.372
81.379
81.386
81.393
81.401
81.408
81.415
81.422
81.430
81.437
81.444
81.451
81.459
81.466
81.473
81.480
81.488
81.495
81.502
81.509
81.517
81.524
81.531
81.538
81.546
81.553
81.560
81.567
81.575
81.582
81.589
81.597
81.604
81.611
81.618
81.626
81.633
81.640
81.647

89.463
89.470
89.478
89.485
89.492
89.499
89.507
89.514
89.521
89.528
89.536
89.543
89.550
89.557
89.565
89.572
89.579
89.586
89.594
89.601
89.608
89.615
89.623
89.630
89.637
89.644
89.652
89.659
89.666
89.673
89.681
89.688
89.695
89.702
89.710
89.717
89.724
89.731
89.739
89.746
89.753
89.761
89.768
89.775
89.782
89.790
89.797
89.804
89.811
89.819
89.826
89.833
89.840
89.848
89.855
89.862
89.869
89.877
89.884
89.891
89.898
89.906
89.913
89.920
89.927
89.935
89.942
89.949
89.956
89.964
89.971
89.978
89.985
89.993
90.000
90.007
90.015
90.022
90.029
90.036
90.044
90.051
90.058
90.065
90.073
90.080
90.087
90.094
90.102
90.109
90.116
90.123
90.131
90.138
90.145
90.152
90.160
90.167
90.174
90.181
90.189
90.196
90.203
90.210
90.218
90.225
90.232
90.239
90.247
90.254
90.261
90.269
90.276
90.283
90.290
90.298
90.305
90.312
90.319
90.327
90.334
90.341
90.348
90.356
90.363
90.370
90.377
90.385
90.392
90.399
90.406
90.414
90.421
90.428
90.435
90.443
90.450
90.457
90.464
90.472
90.479
90.486
90.493

98.374
98.382
98.389
98.396
98.403
98.411
98.418
98.425
98.433
98.440
98.447
98.454
98.462
98.469
98.476
98.483
98.491
98.498
98.505
98.512
98.520
98.527
98.534
98.541
98.549
98.556
98.563
98.570
98.578
98.585
98.592
98.599
98.607
98.614
98.621
98.628
98.636
98.643
98.650
98.657
98.665
98.672
98.679
98.687
98.694
98.701
98.708
98.716
98.723
98.730
98.737
98.745
98.752
98.759
98.766
98.774
98.781
98.788
98.795
98.803
98.810
98.817
98.824
98.832
98.839
98.846
98.853
98.861
98.868
98.875
98.882
98.890
98.897
98.904
98.911
98.919
98.926
98.933
98.940
98.948
98.955
98.962
98.970
98.977
98.984
98.991
98.999
99.006
99.013
99.020
99.028
99.035
99.042
99.049
99.057
99.064
99.071
99.078
99.086
99.093
99.100
99.107
99.115
99.122
99.129
99.136
99.144
99.151
99.158
99.165
99.173
99.180
99.187
99.194
99.202
99.209
99.216
99.224
99.231
99.238
99.245
99.253
99.260
99.267
99.274
99.282
99.289
99.296
99.303
99.311
99.318
99.325
99.332
99.340
99.347
99.354
99.361
99.369
99.376
99.383
99.390
99.398
99.405

8.055
8.062
8.070
8.077
8.084
8.091
8.099
8.106
8.113
8.120
8.128
8.135
8.142
8.149
8.157
8.164
8.171
8.179
8.186
8.193
8.200
8.208
8.215
8.222
8.229
8.237
8.244
8.251
8.258
8.266
8.273
8.280
8.287
8.295
8.302
8.309
8.316
8.324
8.331
8.338
8.345
8.353
8.360
8.367
8.374
8.382
8.389
8.396
8.403
8.411
8.418
8.425
8.433
8.440
8.447
8.454
8.462
8.469
8.476
8.483
8.491
8.498
8.505
8.512
8.520
8.527
8.534
8.541
8.549
8.556
8.563
8.570
8.578
8.585
8.592
8.599
8.607
8.614
8.621
8.628
8.636
8.643
8.650
8.657
8.665
8.672
8.679
8.687
8.694
8.701
8.708
8.716
8.723
8.730
8.737
8.745
8.752
8.759
8.766
8.774
8.781
8.788
8.795
8.803
8.810
8.817
8.824
8.832
8.839
8.846
8.853
8.861
8.868
8.875
8.882
8.890
8.897
8.904
8.911
8.919
8.926
8.933
8.940
8.948
8.955
8.962
8.970
8.977
8.984
8.991
8.999
9.006
9.013
9.020
9.028
9.035
9.042
9.049
9.057
9.064
9.071
9.078
9.086
9.093
9.100
9.107
9.115
9.122
9.129
9.136
9.144
9.151
9.158
9.165
9.173
9.180
9.187
9.194
9.202
9.209
9.216
9.224
9.231
9.238
9.245
9.253
9.26

17.228
17.235
17.242
17.250
17.257
17.264
17.271
17.279
17.286
17.293
17.300
17.308
17.315
17.322
17.329
17.337
17.344
17.351
17.358
17.366
17.373
17.380
17.388
17.395
17.402
17.409
17.417
17.424
17.431
17.438
17.446
17.453
17.460
17.467
17.475
17.482
17.489
17.496
17.504
17.511
17.518
17.525
17.533
17.540
17.547
17.554
17.562
17.569
17.576
17.583
17.591
17.598
17.605
17.612
17.620
17.627
17.634
17.642
17.649
17.656
17.663
17.671
17.678
17.685
17.692
17.700
17.707
17.714
17.721
17.729
17.736
17.743
17.750
17.758
17.765
17.772
17.779
17.787
17.794
17.801
17.808
17.816
17.823
17.830
17.837
17.845
17.852
17.859
17.866
17.874
17.881
17.888
17.896
17.903
17.910
17.917
17.925
17.932
17.939
17.946
17.954
17.961
17.968
17.975
17.983
17.990
17.997
18.004
18.012
18.019
18.026
18.033
18.041
18.048
18.055
18.062
18.070
18.077
18.084
18.091
18.099
18.106
18.113
18.120
18.128
18.135
18.142
18.149
18.157
18.164
18.171
18.179
18.186
18.193
18.200
18.208
18.215
18.222
18.229
18.237
18.244
18.251
18.258

25.747
25.755
25.762
25.769
25.776
25.784
25.791
25.798
25.806
25.813
25.820
25.827
25.835
25.842
25.849
25.856
25.864
25.871
25.878
25.885
25.893
25.900
25.907
25.914
25.922
25.929
25.936
25.943
25.951
25.958
25.965
25.972
25.980
25.987
25.994
26.001
26.009
26.016
26.023
26.030
26.038
26.045
26.052
26.060
26.067
26.074
26.081
26.089
26.096
26.103
26.110
26.118
26.125
26.132
26.139
26.147
26.154
26.161
26.168
26.176
26.183
26.190
26.197
26.205
26.212
26.219
26.226
26.234
26.241
26.248
26.255
26.263
26.270
26.277
26.284
26.292
26.299
26.306
26.313
26.321
26.328
26.335
26.343
26.350
26.357
26.364
26.372
26.379
26.386
26.393
26.401
26.408
26.415
26.422
26.430
26.437
26.444
26.451
26.459
26.466
26.473
26.480
26.488
26.495
26.502
26.509
26.517
26.524
26.531
26.538
26.546
26.553
26.560
26.567
26.575
26.582
26.589
26.597
26.604
26.611
26.618
26.626
26.633
26.640
26.647
26.655
26.662
26.669
26.676
26.684
26.691
26.698
26.705
26.713
26.720
26.727
26.734
26.742
26.749
26.756
26.763
26.771
26.778

34.528
34.536
34.543
34.550
34.557
34.565
34.572
34.579
34.586
34.594
34.601
34.608
34.615
34.623
34.630
34.637
34.644
34.652
34.659
34.666
34.673
34.681
34.688
34.695
34.702
34.710
34.717
34.724
34.731
34.739
34.746
34.753
34.761
34.768
34.775
34.782
34.790
34.797
34.804
34.811
34.819
34.826
34.833
34.840
34.848
34.855
34.862
34.869
34.877
34.884
34.891
34.898
34.906
34.913
34.920
34.927
34.935
34.942
34.949
34.956
34.964
34.971
34.978
34.985
34.993
35.000
35.007
35.015
35.022
35.029
35.036
35.044
35.051
35.058
35.065
35.073
35.080
35.087
35.094
35.102
35.109
35.116
35.123
35.131
35.138
35.145
35.152
35.160
35.167
35.174
35.181
35.189
35.196
35.203
35.210
35.218
35.225
35.232
35.239
35.247
35.254
35.261
35.269
35.276
35.283
35.290
35.298
35.305
35.312
35.319
35.327
35.334
35.341
35.348
35.356
35.363
35.370
35.377
35.385
35.392
35.399
35.406
35.414
35.421
35.428
35.435
35.443
35.450
35.457
35.464
35.472
35.479
35.486
35.493
35.501
35.508
35.515
35.522
35.530
35.537
35.544
35.552
35.559

43.389
43.396
43.403
43.411
43.418
43.425
43.433
43.440
43.447
43.454
43.462
43.469
43.476
43.483
43.491
43.498
43.505
43.512
43.520
43.527
43.534
43.541
43.549
43.556
43.563
43.570
43.578
43.585
43.592
43.599
43.607
43.614
43.621
43.628
43.636
43.643
43.650
43.657
43.665
43.672
43.679
43.687
43.694
43.701
43.708
43.716
43.723
43.730
43.737
43.745
43.752
43.759
43.766
43.774
43.781
43.788
43.795
43.803
43.810
43.817
43.824
43.832
43.839
43.846
43.853
43.861
43.868
43.875
43.882
43.890
43.897
43.904
43.911
43.919
43.926
43.933
43.940
43.948
43.955
43.962
43.970
43.977
43.984
43.991
43.999
44.006
44.013
44.020
44.028
44.035
44.042
44.049
44.057
44.064
44.071
44.078
44.086
44.093
44.100
44.107
44.115
44.122
44.129
44.136
44.144
44.151
44.158
44.165
44.173
44.180
44.187
44.194
44.202
44.209
44.216
44.224
44.231
44.238
44.245
44.253
44.260
44.267
44.274
44.282
44.289
44.296
44.303
44.311
44.318
44.325
44.332
44.340
44.347
44.354
44.361
44.369
44.376
44.383
44.390
44.398
44.405
44.412
44.419

52.300
52.308
52.315
52.322
52.329
52.337
52.344
52.351
52.358
52.366
52.373
52.380
52.388
52.395
52.402
52.409
52.417
52.424
52.431
52.438
52.446
52.453
52.460
52.467
52.475
52.482
52.489
52.496
52.504
52.511
52.518
52.525
52.533
52.540
52.547
52.554
52.562
52.569
52.576
52.583
52.591
52.598
52.605
52.612
52.620
52.627
52.634
52.642
52.649
52.656
52.663
52.671
52.678
52.685
52.692
52.700
52.707
52.714
52.721
52.729
52.736
52.743
52.750
52.758
52.765
52.772
52.779
52.787
52.794
52.801
52.808
52.816
52.823
52.830
52.837
52.845
52.852
52.859
52.866
52.874
52.881
52.888
52.896
52.903
52.910
52.917
52.925
52.932
52.939
52.946
52.954
52.961
52.968
52.975
52.983
52.990
52.997
53.004
53.012
53.019
53.026
53.033
53.041
53.048
53.055
53.062
53.070
53.077
53.084
53.091
53.099
53.106
53.113
53.120
53.128
53.135
53.142
53.149
53.157
53.164
53.171
53.179
53.186
53.193
53.200
53.208
53.215
53.222
53.229
53.237
53.244
53.251
53.258
53.266
53.273
53.280
53.287
53.295
53.302
53.309
53.316
53.324
53.331

60.929
60.936
60.943
60.951
60.958
60.965
60.972
60.980
60.987
60.994
61.001
61.009
61.016
61.023
61.030
61.038
61.045
61.052
61.060
61.067
61.074
61.081
61.089
61.096
61.103
61.110
61.118
61.125
61.132
61.139
61.147
61.154
61.161
61.168
61.176
61.183
61.190
61.197
61.205
61.212
61.219
61.226
61.234
61.241
61.248
61.255
61.263
61.270
61.277
61.284
61.292
61.299
61.306
61.313
61.321
61.328
61.335
61.343
61.350
61.357
61.364
61.372
61.379
61.386
61.393
61.401
61.408
61.415
61.422
61.430
61.437
61.444
61.451
61.459
61.466
61.473
61.480
61.488
61.495
61.502
61.509
61.517
61.524
61.531
61.538
61.546
61.553
61.560
61.567
61.575
61.582
61.589
61.597
61.604
61.611
61.618
61.626
61.633
61.640
61.647
61.655
61.662
61.669
61.676
61.684
61.691
61.698
61.705
61.713
61.720
61.727
61.734
61.742
61.749
61.756
61.763
61.771
61.778
61.785
61.792
61.800
61.807
61.814
61.821
61.829
61.836
61.843
61.851
61.858
61.865
61.872
61.880
61.887
61.894
61.901
61.909
61.916
61.923
61.930
61.938
61.945
61.952
61.959

69.855
69.862
69.869
69.877
69.884
69.891
69.898
69.906
69.913
69.920
69.927
69.935
69.942
69.949
69.956
69.964
69.971
69.978
69.985
69.993
70.000
70.007
70.015
70.022
70.029
70.036
70.044
70.051
70.058
70.065
70.073
70.080
70.087
70.094
70.102
70.109
70.116
70.123
70.131
70.138
70.145
70.152
70.160
70.167
70.174
70.181
70.189
70.196
70.203
70.210
70.218
70.225
70.232
70.239
70.247
70.254
70.261
70.269
70.276
70.283
70.290
70.298
70.305
70.312
70.319
70.327
70.334
70.341
70.348
70.356
70.363
70.370
70.377
70.385
70.392
70.399
70.406
70.414
70.421
70.428
70.435
70.443
70.450
70.457
70.464
70.472
70.479
70.486
70.493
70.501
70.508
70.515
70.522
70.530
70.537
70.544
70.552
70.559
70.566
70.573
70.581
70.588
70.595
70.602
70.610
70.617
70.624
70.631
70.639
70.646
70.653
70.660
70.668
70.675
70.682
70.689
70.697
70.704
70.711
70.718
70.726
70.733
70.740
70.747
70.755
70.762
70.769
70.776
70.784
70.791
70.798
70.806
70.813
70.820
70.827
70.835
70.842
70.849
70.856
70.864
70.871
70.878
70.885

78.752
78.759
78.766
78.774
78.781
78.788
78.795
78.803
78.810
78.817
78.824
78.832
78.839
78.846
78.853
78.861
78.868
78.875
78.882
78.890
78.897
78.904
78.911
78.919
78.926
78.933
78.940
78.948
78.955
78.962
78.970
78.977
78.984
78.991
78.999
79.006
79.013
79.020
79.028
79.035
79.042
79.049
79.057
79.064
79.071
79.078
79.086
79.093
79.100
79.107
79.115
79.122
79.129
79.136
79.144
79.151
79.158
79.165
79.173
79.180
79.187
79.194
79.202
79.209
79.216
79.224
79.231
79.238
79.245
79.253
79.260
79.267
79.274
79.282
79.289
79.296
79.303
79.311
79.318
79.325
79.332
79.340
79.347
79.354
79.361
79.369
79.376
79.383
79.390
79.398
79.405
79.412
79.419
79.427
79.434
79.441
79.448
79.456
79.463
79.470
79.478
79.485
79.492
79.499
79.507
79.514
79.521
79.528
79.536
79.543
79.550
79.557
79.565
79.572
79.579
79.586
79.594
79.601
79.608
79.615
79.623
79.630
79.637
79.644
79.652
79.659
79.666
79.673
79.681
79.688
79.695
79.702
79.710
79.717
79.724
79.731
79.739
79.746
79.753
79.761
79.768
79.775
79.782

87.271
87.279
87.286
87.293
87.300
87.308
87.315
87.322
87.329
87.337
87.344
87.351
87.358
87.366
87.373
87.380
87.388
87.395
87.402
87.409
87.417
87.424
87.431
87.438
87.446
87.453
87.460
87.467
87.475
87.482
87.489
87.496
87.504
87.511
87.518
87.525
87.533
87.540
87.547
87.554
87.562
87.569
87.576
87.583
87.591
87.598
87.605
87.612
87.620
87.627
87.634
87.642
87.649
87.656
87.663
87.671
87.678
87.685
87.692
87.700
87.707
87.714
87.721
87.729
87.736
87.743
87.750
87.758
87.765
87.772
87.779
87.787
87.794
87.801
87.808
87.816
87.823
87.830
87.837
87.845
87.852
87.859
87.866
87.874
87.881
87.888
87.896
87.903
87.910
87.917
87.925
87.932
87.939
87.946
87.954
87.961
87.968
87.975
87.983
87.990
87.997
88.004
88.012
88.019
88.026
88.033
88.041
88.048
88.055
88.062
88.070
88.077
88.084
88.091
88.099
88.106
88.113
88.120
88.128
88.135
88.142
88.149
88.157
88.164
88.171
88.179
88.186
88.193
88.200
88.208
88.215
88.222
88.229
88.237
88.244
88.251
88.258
88.266
88.273
88.280
88.287
88.295
88.302

96.110
96.118
96.125
96.132
96.139
96.147
96.154
96.161
96.168
96.176
96.183
96.190
96.197
96.205
96.212
96.219
96.226
96.234
96.241
96.248
96.255
96.263
96.270
96.277
96.284
96.292
96.299
96.306
96.313
96.321
96.328
96.335
96.343
96.350
96.357
96.364
96.372
96.379
96.386
96.393
96.401
96.408
96.415
96.422
96.430
96.437
96.444
96.451
96.459
96.466
96.473
96.480
96.488
96.495
96.502
96.509
96.517
96.524
96.531
96.538
96.546
96.553
96.560
96.567
96.575
96.582
96.589
96.597
96.604
96.611
96.618
96.626
96.633
96.640
96.647
96.655
96.662
96.669
96.676
96.684
96.691
96.698
96.705
96.713
96.720
96.727
96.734
96.742
96.749
96.756
96.763
96.771
96.778
96.785
96.792
96.800
96.807
96.814
96.821
96.829
96.836
96.843
96.851
96.858
96.865
96.872
96.880
96.887
96.894
96.901
96.909
96.916
96.923
96.930
96.938
96.945
96.952
96.959
96.967
96.974
96.981
96.988
96.996
97.003
97.010
97.017
97.025
97.032
97.039
97.046
97.054
97.061
97.068
97.075
97.083
97.090
97.097
97.104
97.112
97.119
97.126
97.134
97.141

## Building Data Generators

In [11]:
from keras.preprocessing.image import ImageDataGenerator

In [12]:
train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   validation_split=0.4,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)
train_set = train_datagen.flow_from_directory(TRAINING_GENERATOR_PATH,
                                                 target_size = (64, 64),
                                                 batch_size =32,
                                                 class_mode = 'binary',
                                                 subset='training')
validation_set = train_datagen.flow_from_directory(TRAINING_GENERATOR_PATH,
                                                 target_size = (64, 64),
                                                 batch_size =32,
                                                 class_mode = 'binary',
                                                 subset='validation')
test_set = train_datagen.flow_from_directory(TESTING_GENERATOR_PATH,
                                                 target_size = (64, 64),
                                                 batch_size =32,
                                                 class_mode = 'binary',
                                                 subset='validation')

Found 11568 images belonging to 2 classes.
Found 7710 images belonging to 2 classes.
Found 3312 images belonging to 2 classes.


# Modeling

In [ ]:
from keras.models import Sequential
from keras.layers import Conv2D
from keras.layers import MaxPooling2D
from keras.layers import Flatten
from keras.layers import Dense
from keras.layers import Dropout
from keras.optimizers import SGD

classifier = Sequential()

# Step 1 - Convolution
classifier.add(Conv2D(32, (3, 3), input_shape = (64, 64, 3), activation = 'relu'))

# Step 2 - Pooling
classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Adding a second convolutional layer
classifier.add(Conv2D(32, (3, 3), activation = 'relu'))
classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Step 3 - Flattening
classifier.add(Flatten())

# Step 4 - Full connection
classifier.add(Dense(units = 128, activation = 'relu'))
classifier.add(Dropout(rate=0.5))
classifier.add(Dense(units = 1, activation = 'sigmoid'))

# # Compiling the CNN

classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

## Fitting the classifier

**WARNING** : Takes several hours to execute

In [12]:
classifier.fit_generator(train_set,
                           steps_per_epoch=363,
                           epochs=10,
                           validation_data=validation_set,
                           validation_steps=242)

Instructions for updating:
Use tf.cast instead.
Epoch 1/10
363/363 [==============================] - 51s 139ms/step - loss: 0.5440 - acc: 0.7416 - val_loss: 0.3721 - val_acc: 0.8589
Epoch 2/10
363/363 [==============================] - 47s 130ms/step - loss: 0.2922 - acc: 0.8986 - val_loss: 0.3079 - val_acc: 0.9084
Epoch 3/10
363/363 [==============================] - 47s 128ms/step - loss: 0.2322 - acc: 0.9221 - val_loss: 0.2204 - val_acc: 0.8897
Epoch 4/10
363/363 [==============================] - 50s 137ms/step - loss: 0.2222 - acc: 0.9172 - val_loss: 0.1970 - val_acc: 0.9097
Epoch 5/10
363/363 [==============================] - 51s 142ms/step - loss: 0.2090 - acc: 0.9299 - val_loss: 0.2196 - val_acc: 0.8893
Epoch 6/10
363/363 [==============================] - 52s 144ms/step - loss: 0.1955 - acc: 0.9347 - val_loss: 0.1842 - val_acc: 0.9283
Epoch 7/10
363/363 [==============================] - 48s 134ms/step - loss: 0.1897 - acc: 0.9397 - val_loss: 0.1952 - val_acc: 0.9192
Epoch 8

saving the model

In [33]:
model_json = classifier.to_json()
with open("model.json", "w") as json_file:
    json_file.write(model_json)

# serialize weights to HDF5
classifier.save_weights("model.h5")
print("Saved model to disk")

Saved model to disk


Loading

In [34]:
from keras.models import model_from_json
# load json and create model
json_file = open('model.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
loaded_model = model_from_json(loaded_model_json)
# load weights into new model
loaded_model.load_weights("model.h5")
print("Loaded model from disk")

Loaded model from disk


## Evaluating the model on the test set
Testing the model on a data that it hasn't seen before

In [39]:
loaded_model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])
score = loaded_model.evaluate_generator(generator=test_set,steps=102)
print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))

acc: 94.45%


# Using a pre-built model

In [14]:
train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   validation_split=0.4,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)
train_set = train_datagen.flow_from_directory(TRAINING_GENERATOR_PATH,
                                                 target_size = (224,224),
                                                 batch_size =32,
                                                 class_mode = 'binary',
                                                 subset='training')
validation_set = train_datagen.flow_from_directory(TRAINING_GENERATOR_PATH,
                                                 target_size = (224,224),
                                                 batch_size =32,
                                                 class_mode = 'binary',
                                                 subset='validation')
test_set = train_datagen.flow_from_directory(TESTING_GENERATOR_PATH,
                                                 target_size = (224,224),
                                                 batch_size =32,
                                                 class_mode = 'binary',
                                                 subset='validation')

Found 11568 images belonging to 2 classes.
Found 7710 images belonging to 2 classes.
Found 3312 images belonging to 2 classes.


In [1]:
import keras
vgg16_model = keras.applications.vgg16.VGG16()

Using TensorFlow backend.


Instructions for updating:
Colocations handled automatically by placer.


In [2]:
from keras.models import Sequential
model = Sequential()
i = 0
for layer in vgg16_model.layers:
    i +=1
    if i < 23 :
        model.add(layer)
        
for layer in model.layers:
    layer.trainable=False
    
    
from keras.layers import Dense
model.add(Dense(1,activation = 'softmax'))

from keras.optimizers import Adam
model.compile(Adam(lr=0.001), loss='binary_crossentropy', metrics=['accuracy'])


In [15]:
model.fit_generator(train_set,
                           steps_per_epoch=363,
                           epochs=10,
                           validation_data=validation_set,
                           validation_steps=242,
                           verbose=1)


Epoch 1/10
363/363 [==============================] - 6133s 17s/step - loss: 7.9657 - acc: 0.5003 - val_loss: 7.9650 - val_acc: 0.5004
Epoch 2/10
363/363 [==============================] - 6150s 17s/step - loss: 7.9492 - acc: 0.5014 - val_loss: 7.9630 - val_acc: 0.5005
Epoch 3/10
363/363 [==============================] - 6160s 17s/step - loss: 7.9671 - acc: 0.5003 - val_loss: 7.9712 - val_acc: 0.5000
Epoch 4/10
363/363 [==============================] - 6151s 17s/step - loss: 7.9822 - acc: 0.4993 - val_loss: 7.9444 - val_acc: 0.5017
Epoch 5/10
363/363 [==============================] - 6166s 17s/step - loss: 7.9506 - acc: 0.5013 - val_loss: 7.9774 - val_acc: 0.4996
Epoch 6/10
363/363 [==============================] - 6137s 17s/step - loss: 7.9602 - acc: 0.5007 - val_loss: 7.9609 - val_acc: 0.5006
Epoch 7/10
363/363 [==============================] - 6133s 17s/step - loss: 7.9657 - acc: 0.5003 - val_loss: 7.9485 - val_acc: 0.5014
Epoch 8/10
363/363 [==============================] - 6

## Evaluating the model on the test set

In [17]:
score = model.evaluate_generator(generator=test_set,steps=102, verbose=1)
print("%s: %.2f%%" % (model.metrics_names[1], score[1]*100))

102/102 [==============================] - 1038s 10s/step
acc: 50.15%


# Summary

- The manually built model performed well, with an accuracy of 94.45% on the test set
- The prebuilt model used (vgg16) didn't fit well, the final accuracy on the test set is ~50%, it'll be a good idea to use a different one, but as long as the manually constructed moel performed sufficiently well, there's no need for that